load-balancer/src/main.rs

```diff
@@ -1,4 +1,166 @@
+mod load_balancer;
+
+use std::env;
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+use bytes::{Buf, Bytes};
+use http_body_util::{BodyExt, Full};
+use hyper::server::conn::http1;
+use hyper::service::service_fn;
+use hyper::Uri;
+use hyper::{body::Incoming as IncomingBody, header, Method, Request, Response, StatusCode};
+use hyper_util::rt::TokioIo;
+use load_balancer::{BalancingAlgorithm, LoadBalancer, Server};
+use tokio::net::{TcpListener, TcpStream};
+use tokio::sync::RwLock;
+
+type GenericError = Box<dyn std::error::Error + Send + Sync>;
+type Result<T> = std::result::Result<T, GenericError>;
+type BoxBody = http_body_util::combinators::BoxBody<Bytes, hyper::Error>;
+
 #[tokio::main]
-async fn main(){
+async fn main() -> Result<()> {
+    let load_balancer = create_load_balancer().unwrap();
+    let load_balancer = Arc::new(RwLock::new(load_balancer));
+
+    let port = env::var("PORT")
+        .unwrap_or_else(|_| "80".to_string())
+        .parse::<u16>()?;
+    let addr = SocketAddr::from(([0, 0, 0, 0], port));
+    let listener = TcpListener::bind(addr).await.map_err(|e| e.to_string())?;
+    println!("Listening on http://{}", addr);
+
+    loop {
+        let (stream, _) = listener.accept().await.map_err(|e| e.to_string())?;
+        let io = TokioIo::new(stream);
+        let load_balancer_clone = load_balancer.clone();
+
+        tokio::task::spawn(async move {
+            let service = service_fn(move |req| handle_request(req, load_balancer_clone.clone()));
+
+            if let Err(err) = http1::Builder::new().serve_connection(io, service).await {
+                println!("Failed to serve connection: {:?}", err);
+            }
+        });
+    }
+}
+
+fn create_load_balancer() -> Result<LoadBalancer> {
+    let servers = vec![
+        Server::new("127.0.0.1:3000".to_string())?,
+        Server::new("127.0.0.1:3001".to_string())?,
+        Server::new("127.0.0.1:3002".to_string())?,
+    ];
+
+    let lb = LoadBalancer::new(servers)?;
+    Ok(lb)
+}
+
+async fn handle_request(
+    req: Request<IncomingBody>,
+    lb: Arc<RwLock<LoadBalancer>>,
+) -> Result<Response<BoxBody>> {
+    match (req.method(), req.uri().path()) {
+        (&Method::POST, "/algo") => change_algo(req, lb).await,
```

Comment:
Guess this is only for demo purpose of switching the algorithm on the fly.

Reply:
Yes, for it is for demo.